File: scripts/algorithms/M/Maximum Non Negative Product in a Matrix/Maximum Non Negative Product in a Matrix.js
var maxProductPath = function(grid) {
    const R = grid.length, C = grid[0].length;
    if (R === 0 || C === 0)
        return -1;
    
    const mat = [...Array(R)].map(() => [...Array(C)].map(() => new Array(2)));
    
    mat[0][0] = [grid[0][0], grid[0][0]];
    for (let i = 1; i < R; i++)
        mat[i][0] = [mat[i-1][0][0]*grid[i][0], mat[i-1][0][1]*grid[i][0]];
    
    for (let i = 1; i < C; i++)
        mat[0][i] = [mat[0][i-1][0]*grid[0][i], mat[0][i-1][1]*grid[0][i]];
    
    for (let i = 1; i < R; i++) {
        for (let j = 1; j < C; j++) {
            const max = Math.max(mat[i-1][j][0], mat[i][j-1][0]),
                  min = Math.min(mat[i-1][j][1], mat[i][j-1][1]);
            if (grid[i][j] >= 0)
                mat[i][j] = [max*grid[i][j], min*grid[i][j]];
            else
                mat[i][j] = [min*grid[i][j], max*grid[i][j]];
        }
    }
    
    return mat[R-1][C-1][0] >= 0 ? mat[R-1][C-1][0] % (10**9+7) : -1;
};

File: scripts/algorithms/M/Maximum Non Negative Product in a Matrix/Maximum Non Negative Product in a Matrix.java
// Runtime: 4 ms (Top 41.67%) | Memory: 43.3 MB (Top 29.76%)
class Solution {
    public class Pair{
        long min=Integer.MAX_VALUE,max=Integer.MIN_VALUE;
        Pair(){

        }
        Pair(long min,long max){
            this.min=min;
            this.max=max;
        }
    }
    public int maxProductPath(int[][] grid) {
        Pair[][] dp=new Pair[grid.length][grid[0].length];
        for(int r=grid.length-1;r>=0;r--){
            for(int c=grid[0].length-1;c>=0;c--){
                if(r==grid.length-1 && c==grid[0].length-1){
                    dp[r][c]=new Pair(grid[r][c],grid[r][c]);
                }else{
                    Pair hor=(c==grid[0].length-1)?new Pair():dp[r][c+1];
                    Pair ver=(r==grid.length-1)?new Pair():dp[r+1][c];
                    long min,max;
                    if(grid[r][c]>=0){
                         max=Math.max(hor.max,ver.max);
                         min=Math.min(hor.min,ver.min);
                    }else{
                         min=Math.max(hor.max,ver.max);
                         max=Math.min(hor.min,ver.min);
                    }
                        dp[r][c]=new Pair(min*grid[r][c],max*grid[r][c]);
                }
            }
        }
        int mod=(int)1e9 +7;
        return dp[0][0].max<0?-1:(int)(dp[0][0].max%mod);
    }
}